File: nmf3d_mat/inv_expansion_coeffs.m
function varargout=inv_expansion_coeffs(vfile,hfile,wfile,zi,mi,vi,pl,lon,varargin)
%  Inverse expansion coefficients
%  Compute the inverse of Vertical, Fourier and Hough transforms.
%  Recovers the zonal and meridional wind, and geopotential perturbation
%  (from the reference geopotential) in the physical space, for the chosen
%  set of modes. Uses equivalent heights, vertical structure functions
%  Hough functions and complex expansion coefficients.
%
%  vfile: file with the equivalent heights (hk) and vertical structure functions (Gn)
%  hfile: file with the Hough functions (HOUGHs_UVZ)
%  wfile: file with the expansion coefficients (w_nlk)
%  zi:    wavenumber indices, zi=0 (Zonal mean) and zi>0 (Eddies)
%  mi:    meridional indices
%  vi:    vertical indices, vi=0 (Barotropic mode) and vi>0 (Baroclinic modes)
%  pl:    pressure levels (hPa units)
%  lon:   longitudes (deg units)
%  varargin:
%    uvz:   components of wind and mass fields to calculate
%           - default [1 1 1] - calculate the zonal and meridional wind and the geopotential (u, v, z)
%    save, create file [true]
%    format, file format: [nc] or mat
%    attrs, attributes to save [{}]
%    label, filename to save, without extension ['outinv_uvz']
%
%  Returns the reconstructed  zonal and meridional wind, and geopotential perturbation,
%  as well as saved filename if save is true

constants;

uvz=[1,1,1];
save=true; % create file
for i=1:length(varargin)
  if isequal(varargin{i},'uvz')
    uvz=varargin{i+1};
  elseif isequal(varargin{i},'save')
    save=varargin{i+1};
  end
end

% The wavenumber, meridional and vertical indices
ZI = zi+1;
MI = mi;
VI = vi+1;

lon=lon*pi/180; % lon in radians

fprintf(1,' - loading vertical structure functions:\n    %s\n',vfile);
if ends_with(vfile,'.mat')
  ovfile=load(vfile);
  ws0=eval(lower(ovfile.ws0));

  hk    = ovfile.hk;
  Gn    = ovfile.Gn;
  p_new = ovfile.plev;
else
  nc=netcdf.open(vfile);
  ws0 = netcdf.getAtt(nc,netcdf.getConstant('NC_GLOBAL'),'ws0');
  ws0 = eval(lower(ws0));

  hk    = read_nc_var(nc,'hk');
  Gn    = read_nc_var(nc,'Gn');
  p_new = read_nc_var(nc,'plev');

  netcdf.close(nc);
end

fprintf(1,' - loading Hough vector functions:\n    %s\n',hfile);
if ends_with(hfile,'.mat')
  ohfile=load(hfile);
  HOUGHs_UVZ = ohfile.HOUGHs_UVZ;
  lat = ohfile.lat;
else
  nc=netcdf.open(hfile);
  HOUGHs_UVZ=read_nc_var(nc,'HOUGHs_UVZ_real')+1j*read_nc_var(nc,'HOUGHs_UVZ_imag');
  lat=read_nc_var(nc,'lat');
  netcdf.close(nc);
end

fprintf(1,' - loading expansion coefficients:\n    %s\n',wfile);
if ends_with(wfile,'.mat')
  owfile=load(wfile);
  w_nlk = owfile.w_nlk;
else
  nc=netcdf.open(wfile);
  w_nlk=read_nc_var(nc,'w_nlk_real')+1j*read_nc_var(nc,'w_nlk_imag');
  netcdf.close(nc);
end

[three,nN,nmm,nk,nLat]=size(HOUGHs_UVZ);
hk=hk(1:nk);
nT=size(w_nlk,ndims(w_nlk));

if ws0
    hk(1) = 1;
end

% Interpolates vertical structure functions (Gn) for level p with cubic spline
plev=p_new;
plev(1)=const.ps; % surface max pressure, to avoid extrap!
for i=1:nk
  Gn_interpS(i,:) = spline(plev,Gn(i,:),pl*100); % plev pl*100 in Pa
end

% Permuting dimensions of Fourier-Hough transforms for convenience
w_nlk = permute(w_nlk,[2 3 1 4]); % size(w_nlk) = (n,l,k,t)

if max(ZI)>nN
  error('maximum wavenumber was exceeded')
end

if max(VI)>nk
  error('maximum vertical index was exceeded')
end

% variables for the horizontal wind and mass fields expansion into 3-D normal modes
un = zeros(nN, nLat, numel(lon)); % size(u) = (n,lat,lon)
vn = un;
zn = un;

u = zeros(nT, numel(pl), nLat, numel(lon)); % size(u) = (nT,npL,lat,lon)
v = u;
z = u;

select_nlk = zeros(nN,nmm,nk);
select_nlk(ZI,MI,VI) = 1;

fprintf(1,' - computing\n');
for ip=1:numel(pl) % loop over p levels
  Gn = Gn_interpS(:,ip);
  for it=1:nT % loop over times
    A_w_nlk = select_nlk.*squeeze(w_nlk(:,:,:,it));

    % Replicating variables into arrays with matching sizes for
    % elementwise array multiplications (element-by-element). The sizes
    % must mach for dimensions (n,l,k)
    A_h_k = permute(repmat(hk,[1 nN nmm]),[2 3 1]);          % size(A_h_k) = (n,l,k)
    A_Gn = permute(squeeze(repmat(Gn,[1 nN  nmm])),[2 3 1]); % size(A_Gn)  = (n,l,k)

    nW = (0:nN-1).';
    A_Fourier   = repmat(exp(const.I*nW*lon),[1 1 nmm nk]);  % size(A_Fourier)   = (n,lon,l,k)

    for la=1:nLat
      for lo=1:numel(lon)
        % summing over chosen vertical and meridional modes

        un(:,la,lo) = (2*real(squeeze(sum(sum(A_w_nlk .* sqrt(const.g.*A_h_k) .* A_Gn .* squeeze(HOUGHs_UVZ(1,:,:,:,la)) .* squeeze(A_Fourier(:,lo,:,:)),3),2)))).';
        vn(:,la,lo) = (2*real(squeeze(sum(sum(A_w_nlk .* sqrt(const.g.*A_h_k) .* A_Gn .* squeeze(HOUGHs_UVZ(2,:,:,:,la)) .* squeeze(A_Fourier(:,lo,:,:)),3),2)))).';
        zn(:,la,lo) = (2*real(squeeze(sum(sum(A_w_nlk .*      const.g.*A_h_k  .* A_Gn .* squeeze(HOUGHs_UVZ(3,:,:,:,la)) .* squeeze(A_Fourier(:,lo,:,:)),3),2)))).';
      end
    end

    u(it,ip,:,:) = squeeze( sum(un(2:end,:,:),1) + 0.5*un(1,:,:) );
    v(it,ip,:,:) = squeeze( sum(vn(2:end,:,:),1) + 0.5*vn(1,:,:) );
    z(it,ip,:,:) = squeeze( sum(zn(2:end,:,:),1) + 0.5*zn(1,:,:) );
  end
end


data=struct;
data.lon = lon*180/pi;
data.lat = lat;
if uvz(1), data.u=u; end
if uvz(2), data.v=v; end
if uvz(3), data.z=z; end

if save
  % add some input data as attributes:
  attrs=struct;
  j=0;
  for i=1:length(varargin)
    if isequal(varargin{i},'attrs')
      attrs=varargin{i+1};
      j=i;
    end
  end
  attrs.zi  = int32(zi);
  attrs.mi  = int32(mi);
  attrs.vi  = int32(vi);
  attrs.pl  = pl;
  attrs.uvz = int32(uvz);
  if j==0
    varargin{end+1}='attrs';
    varargin{end+1}=attrs;
  else
    varargin{j}=attrs;
  end

  fsave=save_out(data,varargin{:});
  varargout={data,fsave};
else
  varargout={data};
end

end


function tf = ends_with(str, suffix)
  % Return true if the string ends in the specified suffix
  % Matlab >2016b already includes the function endsWith
  n = length(suffix);
  if length(str) < n
    tf =  false;
  else
    tf = strcmp(str(end-n+1:end), suffix);
  end
end


function fsave=save_out(data,varargin)
  label='outinv_uvz';
  format='nc'; % nc or mat
  attrs=struct;

  for i=1:length(varargin)
    if isequal(varargin{i},'label')
      label=varargin{i+1};
    elseif isequal(varargin{i},'format')
      format=varargin{i+1};
    elseif isequal(varargin{i},'attrs')
      attrs=varargin{i+1};
    end
  end

  attrs.platform=computer;
  attrs.environment='matlab';
  attrs.version=version;

  fsave=sprintf('%s.%s',label,format);

  fprintf(1,'saving %s\n',fsave);
  if isequal(format,'mat')
    % update data with attrs:
    f = fieldnames(attrs);
    for i = 1:length(f)
      data.(f{i})=attrs.(f{i});
    end
    save(fsave, '-struct','data');
  elseif isequal(format,'nc')
    save_nc(fsave,data,attrs);
  else
    disp('Unknown format, use nc or mat');
  end

end


function save_nc(fname,data,attrs)
  addu=isfield(data,'u');
  addv=isfield(data,'v');
  addz=isfield(data,'z');

  nc=netcdf.create(fname,'CLOBBER');

  % dimensions:
  if     addu, tmp=data.u;
  elseif addv, tmp=data.v;
  elseif addz, tmp=data.z;
  end

  if addu | addv | addz
    [nT,nLev,nLat,nLon]=size(tmp);

    dim_nT  = netcdf.defDim(nc,'time',nT);
    dim_nP  = netcdf.defDim(nc,'pressure_levels',nLev);
    dim_lat = netcdf.defDim(nc,'lat',nLat);
    dim_lon = netcdf.defDim(nc,'lon',nLon);

    dim=[dim_nT,dim_nP,dim_lat,dim_lon];
    dim=dim(end:-1:1);
  else
    dim_lat = netcdf.defDim(nc,'lat',numel(data.lat));
    dim_lon = netcdf.defDim(nc,'lon',numel(data.lon));
  end

  % variables:
  if addu
    var_u=netcdf.defVar(nc,'u','NC_DOUBLE',dim);
    netcdf.putAtt(nc,var_u,'long_name','reconstructed zonal wind');
  end
  if addv
    var_v=netcdf.defVar(nc,'v','NC_DOUBLE',dim);
    netcdf.putAtt(nc,var_v,'long_name','reconstructed meridional wind');
  end
  if addz
    var_z=netcdf.defVar(nc,'z','NC_DOUBLE',dim);
    netcdf.putAtt(nc,var_z,'long_name','reconstructed geopotential perturbation');
  end

  var_x=netcdf.defVar(nc,'lon','NC_DOUBLE',dim_lon);
  var_y=netcdf.defVar(nc,'lat','NC_DOUBLE',dim_lat);

  netcdf.putAtt(nc,var_x,'long_name','longitude');
  netcdf.putAtt(nc,var_y,'long_name','latitude');

  % global attributes:
  vid = netcdf.getConstant('GLOBAL');
  netcdf.putAtt(nc,vid,'date',datestr(now));
  f=fieldnames(attrs);
  for i=1:length(f)
    netcdf.putAtt(nc,vid,f{i},attrs.(f{i}));
  end

  netcdf.endDef(nc);

  % fill variables:
  if addu, netcdf.putVar(nc,var_u, permute(data.u,ndims(data.u):-1:1) );end
  if addv, netcdf.putVar(nc,var_v, permute(data.v,ndims(data.v):-1:1) );end
  if addz, netcdf.putVar(nc,var_z, permute(data.z,ndims(data.z):-1:1) );end
  netcdf.putVar(nc,var_x, permute(data.lon,ndims(data.lon):-1:1) );
  netcdf.putVar(nc,var_y, permute(data.lat,ndims(data.lat):-1:1) );
  netcdf.close(nc)
end

File: nmf3d/inv_expansion_coeffs.py
from . import constants as const
import numpy as np
import scipy.interpolate
import netCDF4

dType='d'
cType='complex128'

def calc(vfile,hfile,wfile,zi,mi,vi,pl,lon,**kargs):
  '''
  Inverse expansion coefficients
  Compute the inverse of Vertical, Fourier and Hough transforms.
  Recovers the zonal and meridional wind, and geopotential perturbation
  (from the reference geopotential) in the physical space, for the chosen
  set of modes. Uses equivalent heights, vertical structure functions
  Hough functions and complex expansion coefficients.

  vfile: file with the equivalent heights (hk) and vertical structure functions (Gn)
  hfile: file with the Hough functions (HOUGHs_UVZ)
  wfile: file with the expansion coefficients (w_nlk)
  zi:    wavenumber indices, zi=0 (Zonal mean) and zi>0 (Eddies)
  mi:    meridional indices
  vi:    vertical indices, vi=0 (Barotropic mode) and vi>0 (Baroclinic modes)
  pl:    pressure levels (hPa units)
  lon:   longitudes (deg units)
  kargs:
    uvz:   components of wind and mass fields to calculate
           - default [1 1 1] - calculate the zonal and meridional wind and the geopotential (u, v, z)
    save, create file [true]
    format, file format: [nc] or npz
    attrs, attributes to save [{}]
    label, filename to save, without extension ['outinv_uvz']

  Returns the reconstructed  zonal and meridional wind, and geopotential perturbation,
  as well as saved filename if save is true
  '''

  uvz  = kargs.get('uvz',[1,1,1])
  save = kargs.get('save',True)

  # The wavenumber, meridional and vertical indices
  ZI = np.asarray(zi)
  MI = np.asarray(mi)-1
  VI = np.asarray(vi)

  pl = np.asarray(pl)
  lon = np.asarray(lon)

  lon=lon*np.pi/180 # lon in radians

  print(' - loading vertical structure functions:\n    %s'%vfile)
  if vfile.endswith('.npz'):
    ovfile=np.load(vfile)
    ws0=eval(ovfile['ws0'][()]) # ws0 is the string True/False

    hk    = ovfile['hk']
    Gn    = ovfile['Gn']
    p_new = ovfile['plev']
  else:
    nc=netCDF4.Dataset(vfile)
    ws0=eval(nc.ws0)

    hk    = nc.variables['hk'][:]
    Gn    = nc.variables['Gn'][:]
    p_new = nc.variables['plev'][:]
    nc.close()


  print(' - loading Hough vector functions:\n    %s'%hfile);
  if hfile.endswith('.npz'):
    ohfile=np.load(hfile)
    HOUGHs_UVZ = ohfile['HOUGHs_UVZ']
    lat = ohfile['lat']
  else:
    nc=netCDF4.Dataset(hfile)
    HOUGHs_UVZ=nc.variables['HOUGHs_UVZ_real'][:]+1j*nc.variables['HOUGHs_UVZ_imag'][:]
    lat=nc.variables['lat'][:]
    nc.close()

  print(' - loading expansion coefficients:\n    %s'%wfile);
  if wfile.endswith('.npz'):
    owfile=np.load(wfile)
    w_nlk = owfile['w_nlk']
  else:
    nc=netCDF4.Dataset(wfile)
    w_nlk=nc.variables['w_nlk_real'][:]+1j*nc.variables['w_nlk_imag'][:]
    nc.close()

  three,nN,nmm,nk,nLat=HOUGHs_UVZ.shape
  hk=hk[:nk]
  nT=w_nlk.shape[-1]

  if ws0: hk[0]=1

  # Interpolates vertical structure functions (Gn) for level p with cubic spline
  plev=p_new
  plev[0]=const.ps # surface max pressure, to avoid extrap!
  Gn_interpS=np.zeros((nk,pl.size),Gn.dtype)
  for i in range(nk):
    Gn_interpS[i] = scipy.interpolate.UnivariateSpline(plev[::-1],Gn[i][::-1],s=0)(pl*100) # plev pl*100 in Pa

  # Permuting dimensions of Fourier-Hough transforms for convenience
  w_nlk = np.transpose(w_nlk,(1,2,0,3)) # % shape w_nlk  = n,l,k,t

  if max(ZI)>nN:
    raise Exception('maximum wavenumber was exceeded')

  if max(VI)>nk:
    raise Exception('maximum vertical index was exceeded')

  # variables for the horizontal wind and mass fields expansion into 3-D normal modes
  un = np.zeros((nN, nLat, lon.size),dtype=dType) # size u  = n,lat,lon
  vn = np.zeros((nN, nLat, lon.size),dtype=dType)
  zn = np.zeros((nN, nLat, lon.size),dtype=dType)

  u = np.zeros((nT,pl.size,nLat,lon.size),dtype=dType) # size u  = nT,npL,lat,lon
  v = np.zeros((nT,pl.size,nLat,lon.size),dtype=dType)
  z = np.zeros((nT,pl.size,nLat,lon.size),dtype=dType)

  select_nlk = np.zeros((nN,nmm,nk))
  ZI,MI,VI=np.meshgrid(ZI,MI,VI)
  select_nlk[ZI,MI,VI]=1

  print(' - computing')
  lon.shape=1,lon.size
  for ip in range(pl.size): # loop over p levels
    Gn = Gn_interpS[:,ip]
    for it in range(nT): # loop over times
      A_w_nlk = select_nlk*w_nlk[:,:,:,it]

      # Replicating variables into arrays with matching sizes for
      # elementwise array multiplications (element-by-element). The sizes
      # must mach for dimensions (n,l,k)
      A_h_k = np.transpose(np.tile(hk[:,np.newaxis,np.newaxis],[1,nN,nmm]),[1,2,0]) # size A_h_k  = n,l,k
      A_Gn  = np.transpose(np.tile(Gn[:,np.newaxis,np.newaxis],[1,nN,nmm]),[1,2,0]) # size A_Gn   = n,l,k

      nW = np.arange(nN)
      nW.shape=nW.size,1

      tmp=np.exp(const.I*np.matmul(nW,lon))
      A_Fourier=np.tile(tmp[:,:,np.newaxis,np.newaxis],[1,1,nmm,nk]) # size A_Fourier = n,lon,l,k

      for la in range(nLat):
        for lo in range(lon.size):
          # summing over chosen vertical and meridional modes
          un[:,la,lo]=(A_w_nlk*np.sqrt(const.g*A_h_k)*A_Gn*HOUGHs_UVZ[0,:,:,:,la]*A_Fourier[:,lo,:,:]).sum(-1).sum(-1).real*2
          vn[:,la,lo]=(A_w_nlk*np.sqrt(const.g*A_h_k)*A_Gn*HOUGHs_UVZ[1,:,:,:,la]*A_Fourier[:,lo,:,:]).sum(-1).sum(-1).real*2
          zn[:,la,lo]=(A_w_nlk*        const.g*A_h_k *A_Gn*HOUGHs_UVZ[2,:,:,:,la]*A_Fourier[:,lo,:,:]).sum(-1).sum(-1).real*2

      u[it,ip]=un[1:].sum(0)+0.5*un[0]
      v[it,ip]=vn[1:].sum(0)+0.5*vn[0]
      z[it,ip]=zn[1:].sum(0)+0.5*zn[0]

  lon.shape=lon.size
  data=dict(lon=lon*180/np.pi,lat=lat)
  if uvz[0]: data['u']=u
  if uvz[1]: data['v']=v
  if uvz[2]: data['z']=z

  if save:
    # add some input data as attributes:
    attrs=kargs.get('attrs',{})
    attrs.update(dict(zi=zi,mi=mi,vi=vi,pl=pl,uvz=uvz))
    kargs['attrs']=attrs

    fsave=save_out(data,**kargs)
    return data,fsave
  else:
    return data


def save_out(data,**kargs):
  label=kargs.get('label','outinv_uvz')
  format=kargs.get('format','nc') # nc or npz
  attrs=kargs.get('attrs',{})

  import platform
  import sys
  import scipy
  attrs['platform']=platform.platform()
  attrs['environment']='python'
  attrs['version']=sys.version.replace('\n','')
  attrs['version_scipy']=scipy.__version__
  attrs['version_numpy']=np.__version__

  fsave='%s.%s'%(label,format)

  print('saving %s'%fsave)
  if format=='npz':
    data.update(attrs)
    np.savez(fsave, **data)
  elif format=='nc':
    save_nc(fsave,data,**attrs)
  else: print('Unknown format, use nc or npz')

  return fsave


def save_nc(fname,data,**attrs):
  import netCDF4
  import os
  if os.path.isfile(fname):
    os.unlink(fname)

  addu='u' in data
  addv='v' in data
  addz='z' in data

  nc=netCDF4.Dataset(fname,'w',file_format='NETCDF4_CLASSIC')

  # dimensions:
  if   addu: tmp=data['u']
  elif addv: tmp=data['v']
  elif addz: tmp=data['z']

  if addu or addv or addz:
    nT,nLev,nLat,nLon=tmp.shape

    nc.createDimension('time',nT)
    nc.createDimension('pressure_levels',nLev)
    nc.createDimension('lat',nLat)
    nc.createDimension('lon',nLon)

    dim='time','pressure_levels','lat','lon'
  else:
    nc.createDimension('lat',data['lat'].size)
    nc.createDimension('lon',data['lon'].size)

  # variables:
  if addu:
    v=nc.createVariable('u',dType,dim)
    v.long_name='reconstructed zonal wind'

  if addv:
    v=nc.createVariable('v',dType,dim)
    v.long_name='reconstructed meridional wind'

  if addz:
    v=nc.createVariable('z',dType,dim)
    v.long_name='reconstructed geopotential perturbation'

  v=nc.createVariable('lon',dType,('lon',))
  v.long_name='longitude'
  v=nc.createVariable('lat',dType,('lat',))
  v.long_name='latitude'

  # global attributes:
  import datetime
  nc.date=datetime.datetime.now().isoformat(' ')
  for k in attrs.keys():
    setattr(nc,k,attrs[k])

  # fill variables:
  if addu: nc.variables['u'][:]=data['u']
  if addv: nc.variables['v'][:]=data['v']
  if addz: nc.variables['z'][:]=data['z']
  nc.variables['lon'][:]=data['lon']
  nc.variables['lat'][:]=data['lat']
  nc.close()
